sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -150,10 +150,15 @@ case class EnsureRequirements(
         //   A: (No_Exchange, 100) <---> B: (Exchange, 120)
         // it's better to pick A and change B to (Exchange, 100) instead of picking B and insert a
         // new shuffle for A.
-        val candidateSpecsWithoutShuffle = candidateSpecs.filter { case (k, _) =>
-          !children(k).isInstanceOf[ShuffleExchangeLike]
+        val (candidateSpecsWithShuffle, candidateSpecsWithoutShuffle) = candidateSpecs.partition {
+          case (k, _) => children(k).isInstanceOf[ShuffleExchangeLike]
         }
-        val finalCandidateSpecs = if (candidateSpecsWithoutShuffle.nonEmpty) {
+        val maxShufflePartNumOpt = candidateSpecsWithShuffle.values.map(_.numPartitions).maxOption
+        val finalCandidateSpecs = if (candidateSpecsWithoutShuffle.nonEmpty &&
+          maxShufflePartNumOpt.forall { n =>
+            n / candidateSpecsWithoutShuffle.values.map(_.numPartitions).max <=
+              conf.coalesceBucketsInJoinMaxBucketRatio
```

Comment:
just for my education, where else do we read this conf?

Reply:
https://github.com/apache/spark/blob/498519ee6bb4b0295d1df005175e4cbcbcb051e3/sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInJoin.scala#L89-L90